gnomad_qc/v4/assessment/logistic_regression.py

```diff
@@ -0,0 +1,179 @@
+    ht3 = ht.filter(
+        (ht.freq_comparison_stats.stat_union.p_value >= 10e-4)
+        & (
+            ht.freq_comparison_stats.stat_union.stat_test_name
+            == "cochran_mantel_haenszel_test"
+        )
+    )
+    ht3 = ht3._filter_partitions(range(5))
+    ht = ht1.union(ht2).union(ht3)
+
+    # make this an interval table
+    ht = ht.annotate(
+        interval=hl.locus_interval(
+            ht.locus.contig, ht.locus.position, ht.locus.position + 1
+        )
+    )
+    ht = ht.key_by(ht.interval)
+    ht = ht.select()
+
+    return ht
+
+
+def densify_union_exomes_genomes(
+    exomes_vds: hl.vds.VariantDataset,
+    genomes_vds: hl.vds.VariantDataset,
+) -> hl.MatrixTable:
+    """
+    Union exomes and genomes MatrixTables.
+
+    :param exomes_vds: VDS of exomes
+    :param genomes_vds: VDS of genomes
+    :return: Unioned MatrixTable
+    """
+    logger.info("Densifying exomes...")
+    exomes_mt = hl.vds.to_dense_mt(exomes_vds)
+    exomes_mt = exomes_mt.annotate_cols(is_genome=False)
+    exomes_mt = exomes_mt.select_entries("GT").select_rows().select_cols("is_genome")
+
+    logger.info("Densifying genomes...")
+    genomes_mt = hl.vds.to_dense_mt(genomes_vds)
+    genomes_mt = genomes_mt.annotate_cols(is_genome=True)
+    genomes_mt = genomes_mt.select_entries("GT").select_rows().select_cols("is_genome")
+
+    logger.info("Unioning exomes and genomes...")
+    mt = exomes_mt.union_cols(genomes_mt)
+    return mt
+
+
+def main(args):
+    """Perform logistic regression on the unioned exomes and genomes MatrixTable."""
+    hl.init(
+        log="/logistic_regression.log",
+        default_reference="GRCh38",
+        tmp_dir="gs://gnomad-tmp-30day",
+    )
+
+    test = args.test
+    overwrite = args.overwrite
+
+    exomes_vds = get_gnomad_v4_vds(
+        release_only=True,
+        split=True,
+    )
+    genomes_vds = get_gnomad_v4_genomes_vds(
+        release_only=True,
+        split=True,
+    )
+
+    ht = release_sites(data_type="joint").ht()
+    # Maybe only filter to variants with a union p-value < 10e-4 later
+
+    if test:
+        logger.info("Filtering to test intervals...")
+        ht = get_test_intervals(ht)
+        ht = ht.checkpoint(hl.utils.new_temp_file("test_intervals", "ht"))
+        exomes_vds = hl.vds.filter_intervals(
+            exomes_vds, ht, split_reference_blocks=True
+        )
+        genomes_vds = hl.vds.filter_intervals(
+            genomes_vds, ht, split_reference_blocks=True
+        )
+
+    mt = densify_union_exomes_genomes(exomes_vds, genomes_vds)
+    mt = mt.checkpoint(
+        hl.utils.new_temp_file("union_exomes_genomes", "mt"),
+    )
+
+    pc_scores = ancestry_pca_scores().ht()
+    mt = mt.annotate_cols(pc=pc_scores[mt.col_key].scores)
+
+    # Perform logistic regression
+    ht = hl.logistic_regression_rows(
+        "firth",
+        y=mt.is_genome,
+        x=mt.GT.n_alt_alleles(),
+        covariates=[1] + [mt.pc[i] for i in range(10)],
```

Comment:
I don't remember how many PCs we used for ancestry assignment off the top of my head, but I would use that number

Reply:
Mike told me it was 10, but I do remember you're exploring until 18,19, I will double check the code.

Reply:
Sorry if I wasnt clear, I wasnt certain it was 10, just thought it may be. Its 20: https://app.zenhub.com/workspaces/gnomad-5f4d127ea61afc001d6be50b/issues/gh/broadinstitute/gnomad_production/496

Reply:
No worries, I put that 10 temporarily because it was just 10 in Julia's code. I changed it in my test.